### ISSUE TITLE
Update base classes to support (partially) generic tensors and batches

### ISSUE THREAD

**Issue by sgreenbury** [alan-turing-institute/auto-cast#49]
Following #47, it could be worthwhile to extend the base classes to support generics instead of specific types.

This would enable the base classes to be used for different data sources that may have different batch structures.

However, it would be helpful to rely on some common data structure within a batch - this would enable default implementations to be used (e.g. for the loss a `targets` or `outputs` could be useful). This partial structure could be facilitated with a `Protocol`.

Something like the below could work well:
```python
# Generic tensor type
TensorT = TypeVar("TensorT")

# Batches must have at least 'inputs' and 'outputs' attributes
class BatchProtocol(Protocol[TensorT]):
    """Protocol for batch-like objects."""

    inputs: TensorT
    outputs: TensorT

# Generic batch type
BatchT = TypeVar("BatchT", bound=BatchProtocol)

@dataclass
class Batch(Generic[TensorT]):
    """A batch in input data space."""
    # inputs and outputs are same type here but could also have different generics
    inputs: TensorT
    outputs: TensorT
    constant_scalars: TensorBC | None
    constant_fields: TensorBSPlusC | None


@dataclass
class EncodedBatch(Generic[TensorT]):
    """A batch after being processed by an Encoder."""

    inputs: TensorT
    outputs: TensorT
    encoded_info: dict[str, Tensor]
```

This could then be used for example in the encoder:
```python
class Encoder(nn.Module, ABC, Generic[BatchT, TensorT]):
    # ...
    def preprocess(self, batch: BatchT) -> BatchT:
        return batch
    def encode(self, batch: BatchT) -> TensorT: ...
```

And then in specific subclasses, e.g.:
```python
class DCEncoder(Encoder[Batch[TensorBTSPlusC], TensorBTSPlusC]):
    """Deep Compressed (DC) encoder module for (Batch, Time, Spatial dims, Channels)"""
```
